README
jp2a
====

![full-check](https://github.com/Talinx/jp2a/workflows/full-check/badge.svg?branch=master)


jp2a is a simple JPEG/PNG to ASCII converter.

The source is available at <https://github.com/Talinx/jp2a>. This project
used to be hosted on SourceForge.

License
-------

Copyright 2006-2016 Christian Stigen Larsen
Copyright 2020 Christoph Raitzig  
Copyright 2022 Ronald Record  
Distributed under the GNU General Public License v2.

You can read the full license in the file COPYING.  jp2a also uses other
libraries.  These licenses can be found in the file LICENSE.

Example: Oswald the lucky rabbit
--------------------------------
                                                                 ......
                                                         .:lOKMMMMMMMMMM0l
                                                    .;dOMMMMMMMMMMMMMMMMMMW.
                                                .ckWMMMMMMMMMMMMMMMMMMMMMMMO
                                             .dXMMMMMMMMMMMMMMMMMMMMMMMMMMO
                                          .xWMMMMMMMMMMMMMMMMMMMMMMMMMMO:.
                                        lXMMMMMMMMMMMMMMMMMMMMMWOOxc'
                    .;lxOOOOOxdl:.    cMMMMMMMMMMMMMMMMWKklc'..
                'llc..   ..:oMMMMMMKxNMMMMMMMMMM0kl:'.
              :0'             cMMMMMMMMMMMMWd'.
            .K:l0              .MMMMMMMMMMl.
           ,K    W              ,MMMMMMMMMMMMN0dl'
           M.    M               MMMMMMMMMMMMMMMMMMKx:.
          cMMK                   WMMMMMMMMMMMMMMMMMMMMMXd'
          NMMM:                  MMMMMMMMMMMXWMMMMMMMMMMMMNd'
          KMMMW                  MMMMMMMMMMM' ;XMMMMMMMMMMMMMWd'
           MMMM                 XMMMMMMMMMMM    dMMMMMMMMMMMMMMMNc
      ''   NMKWl'             .KMMMMMMMMMMM'     :MMMMMMMMMMMMMMMMMd
     KMWMKOk    '           ,lXMMMMMMMMMMM;       .KMMMMMMMMMMMMMMMM0
     KMMMMM.                    .cOMMMMMW'          kMMMMMMMMMMMMMMMMX
      cOXMNXl.            ......   .MWd. .0cc:'      .NMMMMMMMMMMMMMMMk
            .cololoxKWWWXWMMMMMMM:  M   c:. .l0ok      oMMMMMMMMMMMMMMM
                      .'Mo,MMMMNk:lO' :l  .OMMMMl       .OMMMMMMMMMMMMN
                       X0'OkxkxOK;.  :: ,.XMMMMMMO    :O   lXMMMMMMMMk
                        kc,,o0M0.:;c'N  ko;lMMMMMM:'xMMMNWKO,.clxdl:.
                       ;00xkM:d ,   ..     oMMMMMM cMMMMMMMMl
                       c   N0l       .  .l:;;oKMMk  MMMMMMMMM0
                       O   ,.ocOc..l..k'   .:c;'O   0MMMMMMMX.
                      X     N..l  .c           K.   OMMMMMMM
                     .'     c.     :           M .  OMMM0,.;
                    :,   N.;W       ;          O...o .,.
                    d    M  ;X'     :l             :
                    d..':. lcKol,,,,'.:c,,'.cd;'o,'.
                      . ;:,WMMMMMMXx,; M0XXNNWM.
                           MMMMMMMM   .MMMMMMMM
                           MMMMMMMM   .MMMMMMMM.
                          :ONMMMMMM0:':MMMMMMMMM:
                ;c:,'',::lX'  '',''   .MMMMMMMMMMWK:c:
               x.          .lloc;.',lxl..lKMMMMMKl   M
              0           .;,;'......;::c'  ..     ;dl;l'
             ;O:        .d'              .lkk::::;,     'k
             N.'oc'.    M                   :            X
              lo:. ',::xN:                             .x0'
                .,lccccM .ll,.                . ;;:::;c'.0
                       .kl.  .c::lcccccc:cc::;  .''clccl.
                          'lcl:lc::,,'::cc:cc:c,..

The Oswald character is Copyright Walt Disney, and, unfortunately, not in the
public domain.  However, I believe that I can use this picture under the fair
use clause, for demonstrational purposes, especially in low-resolution, ASCII
format. If I am wrong, I will promptly remove it.

Installation
------------

Look in your system's package manager. I don't maintain Windows binaries
anymore.

Docker
------

You can run jp2a using [Docker](https://www.docker.com/):
```
docker run -t --rm -e COLORTERM="$COLORTERM" -v "$(pwd)":/app talinx/jp2a [options] <filenames or URLs>
```
See [hub.docker.com/r/talinx/jp2a](https://hub.docker.com/r/talinx/jp2a) for details.

Building
--------

`jp2a` can be compiled as usual: `./configure && make && make install`

If you want to compile after cloning the git repo (i.e. generate the `configure` script yourself), you need autotools, automake and autoconf-archive. You may need to specify the path to jpeglib. I usually just do

    autoreconf -vi
    ./configure --with-jpeg-prefix=/usr/local \
                --with-curl-config=`which curl-config`
    make -j

Before installing, you may want to check that jp2a works by running `src/jp2a`
on some `.jpg` and `.png` files. You can also use `make check`. When you're ready to install,

    make -j install

Extract the archive and follow the instructions inside 'install.txt' to build and
install it. For a default install, the corresponding `.so` (shared object) files
will be present in `/usr/local/lib/`

Next, run the command `ldconfig` (escalated privileges may be required) to configure dynamic linker run-time bindings. See
its man page for more details. The configure command should work after that.

If configure can't automatically find jpeglib, specify with

    ./configure --with-jpeg-include=PATH --with-jpeg-lib=PATH

or you can just type the prefix

    ./configure --with-jpeg-prefix=/usr/local/

You can also compile jp2a with libcurl, so it can automatically download images
from the net:

    ./configure --enable-curl

This will look for `curl-config` in $PATH, but you can specify it directly with
`--with-curl-config=PATH` if you need to.

### Runtime dependencies
- [jpeglib](https://www.ijg.org/) or [libjpeg-turbo](https://libjpeg-turbo.org/)
- [libpng](http://libpng.org/pub/png/libpng.html)
- [curl](https://curl.haxx.se/) (optional)
- [ncurses](https://invisible-island.net/ncurses/ncurses.html) (optional)

### Build dependencies
- make (e.g. [GNU Make](https://www.gnu.org/software/make/))
- a C compiler like [gcc](https://gcc.gnu.org/)

If you generate the `configure` script yourself (i.e. run `autoreconf`) you also need these:

- autotools
- [automake](https://www.gnu.org/software/automake/)
- [autoconf](https://www.gnu.org/software/autoconf/autoconf.html)
- [autoconf-archive](https://www.gnu.org/software/autoconf-archive/)

Windows
-------

### Cross compiling

You can cross compile.

### WSL

You can use the Windows Subsystem for Linux.

### Compile with `configure`

The `configure` script requires a Bourne shell, so you got to install one (e.g. Bash).
The easiest way is to install [Cygwin](https://cygwin.com/). Make sure to install the used libraries (curl, jpeglib, libpng, ncurses) as well as gcc and make.
Open Cygwin and proceed as usual: `./configure && make && make install`

You can also "cross compile" with Cygwin so that you get a binary that does not depend on Cygwin.

### Compile without `configure`

You need a working jpeglib for Windows and set up include paths yourself.
You will need to manually type out a `config.h` file. Example:

    WIN32
    NDEBUG
    _CONSOLE
    PACKAGE_STRING=\"jp2a x.x.x-win32\"
    PACKAGE_BUGREPORT=\"chris-r@posteo.net\"
    HAVE_STRING_H
    HAVE_STDLIB_H
    FEAT_CURL
    HAVE_CURL_CURL_H
    CURL_STATICLIB
    FEAT_TERMLIB

You need to add jpeg.lib to the linker. If you want libcurl support, you need
that as well. It should be available from https://curl.haxx.se


ChangeLog
CHANGES SINCE 1.0.0 release 1
- Changed Debian packaging ImageMagick dependency

CHANGES SINCE move to any2ascii repository (version 1.0.0 release 1)
- Split jp2a out of Asciiville project into standalone repository 'any2ascii'
- Add 'any2a' command to convert any image format to jpeg then ascii
- Packaging and distribution for Arch, Fedora, Raspberry Pi OS, and Ubuntu

CHANGES SINCE 1.1.1
- Fix images with very high aspect ratio, e.g. 3000:1

CHANGES SINCE 1.1.0
- support for PNG images
- support for HTML (Living Standard) additionally to XHTML
- (optional) support for other character encodings than ASCII

CHANGES SINCE 1.0.9
- support for true color and 256 color palette output for terminals
- support for bash completion

CHANGES SINCE 1.0.8
- Fix images with very small height

CHANGES SINCE 1.0.7
- Moved repository over to GitHub and made some corresponding changes
- Reverted code to version 1.0.6 because of unfinished changes that didn't work.

CHANGES SINCE 1.0.6 (svnversion 460)
- Internal code changes for resizing.
- Renamed `make test' to `make check'.
- Some missing function signatures were added.
- Using C99 keyword `restrict' in some places.

CHANGES SINCE 1.0.5 (svnversion 452)
- Support for GCC 2.95.4 (old C compilers in general)

CHANGES SINCE 1.0.4 (svnversion 445)
- Fixed problem with --invert / --background=light / --background=dark with HTML output.
- Updated configure script from autoconf 2.59 to 2.60

CHANGES SINCE 1.0.3 (svnversion 440)
- HTML output no longer uses the &apos; entity.

CHANGES SINCE 1.0.2 (svnversion 432)
- Added option --html-no-bold to set non-bold output.
- Added option to --fill ANSI background colors.
- Option --html-fill is now same as --fill and will be phased out.
- Default HTML font changed from courier new to courier.
- HTML output always uses <pre> tag.

CHANGES SINCE 1.0.1 (svnversion 427)
- Using span- instead of b-tags in HTML output.

CHANGES SINCE 1.0.0 (svnversion 421)
- Added option --html-title=...
- Added option --html-raw

CHANGES SINCE 0.9.22
- License changed from "GPL v2 or later" to "GPL v2".  This is just because
  I want to review the v3 license before deciding to apply it to jp2a, as it
  might require me to add things to the program.
- Minor code changes (in image.c)
- Distribution now also contains HTML manual pages for systems that do not support UNIX man.
- Version bumped to 1.0.0 (stable)

CHANGES SINCE 0.9.21 (rev. 412)
- Text color output is a bit more colorful.
- HTML-output is now Firefox-friendly.
- Added --html-fill to enable/disable background colors on HTML output.
- Options --background=light or =dark now affects HTML output background.

CHANGES SINCE 0.9.20
- Added color output for text (ANSI colors) and HTML (CSS colors).
- Changed default HTML fontsize from 4 to 8 pt.
- Fixed a bug that produced incorrect output when using --output and --html.
- Fixed a bug that gave incorret --term-fit when vieweing multiple files.

CHANGES SINCE 0.9.19
- A bash script is used for testing now.
- The tests were updated to use --width=78 (the old default) where appropriate.
- Cygwin support was added.
- term.h is now also sought in ncurses/term.h (for Cygwin support).
- The help text was shrunk.

CHANGES SINCE 0.9.18
- Default running mode is now "jp2a --term-fit" if you have --enable-termlib (which is also default).
- New options --term-fit, --term-width, --term-height to automatically fit image to terminal display.
- New options --background=dark and --background=light so you don't have to bother with --invert.
- Much better output quality as all source pixels are now read.
- Configure now accepts --enable-termlib (default) or --disable-termlib.
- Bugfix: In some cases, the last pixel(s) of the source images were not read.
- Will now gracefully exit if jpeg color precision is other than 8 bits per channels.

CHANGES SINCE 0.9.17
- Bugfix:  No more crashes when reading non-JPEG files on Windows.
- New option: --output=file, send output to given file
- New options: -x for shorthand --flipx, and -y for --flipy.
- New options: --red, --green, --blue.
- More natural RGB to Grayscale calculation.
- Updated manuals.
- Cleaner code.
- Faster code (most evident in large images).

CHANGES SINCE 0.9.16
- New feature: Download-support on win32 binary builds.
- Added option to --clear screen before writing each output image.
- Updated tests for win32 cygwin environment (but still incomplete).

CHANGES SINCE 0.9.15 AND 0.9.14 
- New feature:  Now follow HTTP redirects when downloading.
- Cleaner and faster code.
- Cleaner package organization.
- Improved `make test'.
- Binary is larger becase of debug-symbols; remove with `make install-strip'.
- Added HTML-manual for win32 binary distribution.

CHANGES SINCE 0.9.13
- Minor bugfix: --chars now correctly accepts minimum two chars.
- Changed package layout (added src/ and tests/).
- Added `make test' target.

CHANGES SINCE 0.9.12
- Bugfix: Fixed bug when viewing several files (in clear(), yadds wasn't completely cleared)
- Minor code cleanup.

CHANGES SINCE 0.9.11
- Some Linux build failed on this:  curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, NULL);
  This has been removed, and works well for the Linux distro tested on.

CHANGES SINCE 0.9.10
- New feature:  Can now download images from URLs (requires libcurl)
- New feature:  Added --debug (-d) for debug output with libcurl
- configure:    Added --enable-curl and --with-curl-config=PATH

CHANGES SINCE 0.9.9
- Using both --width and --height is equivalent to using --size
- Now correctly calculates width and height if one of --width or --height is used

CHANGES SINCE 0.9.8
- New feature: Added --flipx and --flipy options to flip the output image.
- New feature: Added --html and --html-fontsize for XHTML 1.0 output.
- Bugfix: The bugfix from 0.9.6 sneaked back into the 0.9.8 release, and is now removed.
- Code changes: The code is a little bit cleaner, and faster now for big output dimensions

pkg/mkaur.sh
#!/bin/bash
#
# mkaur.sh - front end for Arch Linux PKGBUILD distribution creation

PKG="any2ascii"
PKG_NAME="any2ascii"
FULLNAME="Ronald Record"
EMAIL="ronaldrecord@gmail.com"
DESTDIR="usr"
ARCH=x86_64
SUDO=sudo
GCI=

have_makepkg=`type -p makepkg`
[ "${have_makepkg}" ] || {
  echo "Arch Linux packaging tools do not appear to be installed"
  echo "Are you on the appropriate Linux system with packaging requirements ?"
  echo "Exiting"
  exit 1
}

arch_arch=`uname -m`
[ "${arch_arch}" == "${ARCH}" ] || ARCH=${arch_arch}

if [ "${__ASC_SRC__}" ]
then
  SRC="${__ASC_SRC__}"
else
  SRC="${HOME}/src/${PKG_NAME}"
fi

[ -f "${SRC}/VERSION" ] || {
  [ -f "/builds/doctorfree/${PKG_NAME}/VERSION" ] || {
    echo "${SRC}/VERSION does not exist. Exiting."
    exit 1
  }
  SRC="/builds/doctorfree/${PKG_NAME}"
  GCI=1
# SUDO=
}

. "${SRC}/VERSION"
PKG_VER=${VERSION}
PKG_REL=${RELEASE}

umask 0022

# Subdirectory in which to create the distribution files
OUT_DIR="dist/${PKG_NAME}_${PKG_VER}"

[ -d "${SRC}" ] || {
    echo "$SRC does not exist or is not a directory. Exiting."
    exit 1
}

cd "${SRC}"

${SUDO} rm -rf ${PKG}
mkdir ${PKG}
cp VERSION ${PKG}/VERSION
cp pkg/aur/PKGBUILD ${PKG}/PKGBUILD
cp pkg/aur/${PKG}.install ${PKG}/${PKG}.install
cp pkg/aur/.SRCINFO ${PKG}/.SRCINFO
cp pkg/aur/makepkg.conf ${PKG}/makepkg.conf

echo "Creating ${PKG_NAME} PKGBUILD distribution archive"
tar cf - ${PKG} | \
    gzip -9 > ${PKG_NAME}-pkgbuild-${PKG_VER}-${PKG_REL}.tar.gz
mv ${PKG_NAME}-pkgbuild-${PKG_VER}-${PKG_REL}.tar.gz ${PKG}

echo "Building ${PKG_NAME}_${PKG_VER} AUR package"
cd "${SRC}/${PKG}"
makepkg --force --log --cleanbuild --noconfirm --syncdeps

# Rename package if necessary
for zstfile in *.zst
do
  [ "${zstfile}" == "*.zst" ] && continue
  newnam=`echo ${zstfile} | sed -e "s/${PKG}-${PKG_VER}-${PKG_REL}/${PKG_NAME}_${PKG_VER}-${PKG_REL}/"`
  [ "${zstfile}" == "${newnam}" ] && continue
  mv ${zstfile} ${newnam}
done

[ "${GCI}" ] || {
  [ -d ../releases ] || mkdir ../releases
  [ -d ../releases/${PKG_VER} ] || mkdir ../releases/${PKG_VER}
  ${SUDO} cp *.zst *.tar.gz ../releases/${PKG_VER}
}


pkg/mkrpm.sh
#!/bin/bash
PKG="any2ascii"
SRC_NAME="any2ascii"
PKG_NAME="any2ascii"
DESTDIR="usr"
SRC=${HOME}/src
SUDO=sudo
GCI=

[ -f "${SRC}/${SRC_NAME}/VERSION" ] || {
  [ -f "/builds/doctorfree/${SRC_NAME}/VERSION" ] || {
    echo "$SRC/$SRC_NAME/VERSION does not exist. Exiting."
    exit 1
  }
  SRC="/builds/doctorfree"
  SUDO=
  GCI=1
}

. "${SRC}/${SRC_NAME}/VERSION"
PKG_VER=${VERSION}
PKG_REL=${RELEASE}

umask 0022

# Subdirectory in which to create the distribution files
OUT_DIR="dist/${PKG_NAME}_${PKG_VER}"

[ -d "${SRC}/${SRC_NAME}" ] || {
    echo "$SRC/$SRC_NAME does not exist or is not a directory. Exiting."
    exit 1
}

cd "${SRC}/${SRC_NAME}"

# Install required development environment tools
PKGS="automake libtool ncurses-devel"
${SUDO} dnf -y groupinstall "Development Tools" "Development Libraries"
${SUDO} dnf -y install ${PKGS} pandoc zip

# Build any2ascii
./build

${SUDO} rm -rf dist
mkdir dist

[ -d ${OUT_DIR} ] && rm -rf ${OUT_DIR}
mkdir ${OUT_DIR}

for dir in "${DESTDIR}" "${DESTDIR}/bin" "${DESTDIR}/share" \
           "${DESTDIR}/share/doc" "${DESTDIR}/share/doc/${PKG}" \
           "${DESTDIR}/share/man" "${DESTDIR}/share/man/man1" \
           "${DESTDIR}/share/${PKG}"
do
    [ -d ${OUT_DIR}/${dir} ] || ${SUDO} mkdir ${OUT_DIR}/${dir}
    ${SUDO} chown root:root ${OUT_DIR}/${dir}
done

# Install any2ascii
${SUDO} cp src/jp2a ${OUT_DIR}/${DESTDIR}/bin/jp2a
${SUDO} cp src/any2a ${OUT_DIR}/${DESTDIR}/bin/any2a
${SUDO} cp man/jp2a.1 ${OUT_DIR}/${DESTDIR}/share/man/man1

${SUDO} cp AUTHORS ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}/AUTHORS
${SUDO} cp BUGS ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}/BUGS
${SUDO} cp COPYING ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}/COPYING
${SUDO} cp README ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}/README
${SUDO} cp copyright ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}
${SUDO} cp ChangeLog ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}
${SUDO} cp README.md ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}
${SUDO} cp LICENSE ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}
${SUDO} cp NEWS ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}
${SUDO} cp VERSION ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}
${SUDO} pandoc -f gfm README.md | ${SUDO} tee ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}/README.html > /dev/null
${SUDO} gzip -9 ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}/ChangeLog

${SUDO} chmod 755 ${OUT_DIR}/${DESTDIR}/bin/* ${OUT_DIR}/${DESTDIR}/bin
${SUDO} chown -R root:root ${OUT_DIR}/${DESTDIR}

echo "Building ${PKG_NAME}_${PKG_VER} rpm package"

[ -d pkg/rpm ] && cp -a pkg/rpm ${OUT_DIR}/rpm
[ -d ${OUT_DIR}/rpm ] || mkdir ${OUT_DIR}/rpm

have_rpm=`type -p rpmbuild`
[ "${have_rpm}" ] || {
  have_yum=`type -p yum`
  if [ "${have_yum}" ]
  then
    ${SUDO} yum -y install rpm-build
  else
    ${SUDO} apt-get update
    export DEBIAN_FRONTEND=noninteractive
    ${SUDO} ln -fs /usr/share/zoneinfo/America/Los_Angeles /etc/localtime
    ${SUDO} apt-get install rpm -y
    ${SUDO} dpkg-reconfigure --frontend noninteractive tzdata
  fi
}

rpmbuild -ba --build-in-place \
   --define "_topdir ${OUT_DIR}" \
   --define "_sourcedir ${OUT_DIR}" \
   --define "_version ${PKG_VER}" \
   --define "_release ${PKG_REL}" \
   --buildroot ${SRC}/${SRC_NAME}/${OUT_DIR}/BUILDROOT \
   ${OUT_DIR}/rpm/${PKG_NAME}.spec

# Rename RPMs if necessary
for rpmfile in ${OUT_DIR}/RPMS/*/*.rpm
do
  [ "${rpmfile}" == "${OUT_DIR}/RPMS/*/*.rpm" ] && continue
  rpmbas=`basename ${rpmfile}`
  rpmdir=`dirname ${rpmfile}`
  newnam=`echo ${rpmbas} | sed -e "s/${PKG_NAME}-${PKG_VER}-${PKG_REL}/${PKG_NAME}_${PKG_VER}-${PKG_REL}/"`
  [ "${rpmbas}" == "${newnam}" ] && continue
  mv ${rpmdir}/${rpmbas} ${rpmdir}/${newnam}
done

${SUDO} cp ${OUT_DIR}/RPMS/*/*.rpm dist

[ "${GCI}" ] || {
    [ -d releases ] || mkdir releases
    [ -d releases/${PKG_VER} ] || mkdir releases/${PKG_VER}
    ${SUDO} cp ${OUT_DIR}/RPMS/*/*.rpm releases/${PKG_VER}
}


pkg/mkdeb.sh
#!/bin/bash
PKG="any2ascii"
SRC_NAME="any2ascii"
PKG_NAME="any2ascii"
DEBFULLNAME="Ronald Record"
DEBEMAIL="ronaldrecord@gmail.com"
DESTDIR="usr"
SRC=${HOME}/src
ARCH=amd64
SUDO=sudo
GCI=

dpkg=`type -p dpkg-deb`
[ "${dpkg}" ] || {
    echo "Debian packaging tools do not appear to be installed on this system"
    echo "Are you on the appropriate Linux system with packaging requirements ?"
    echo "Exiting"
    exit 1
}
dpkg_arch=`dpkg --print-architecture`
[ "${dpkg_arch}" == "${ARCH}" ] || ARCH=${dpkg_arch}

[ -f "${SRC}/${SRC_NAME}/VERSION" ] || {
  [ -f "/builds/doctorfree/${SRC_NAME}/VERSION" ] || {
    echo "$SRC/$SRC_NAME/VERSION does not exist. Exiting."
    exit 1
  }
  SRC="/builds/doctorfree"
  GCI=1
# SUDO=
}

. "${SRC}/${SRC_NAME}/VERSION"
PKG_VER=${VERSION}
PKG_REL=${RELEASE}

umask 0022

# Subdirectory in which to create the distribution files
OUT_DIR="dist/${PKG_NAME}_${PKG_VER}"

[ -d "${SRC}/${SRC_NAME}" ] || {
    echo "$SRC/$SRC_NAME does not exist or is not a directory. Exiting."
    exit 1
}

cd "${SRC}/${SRC_NAME}"

# Install required development environment tools
${SUDO} apt install autotools-dev autoconf libtool build-essential libncurses-dev

# Build any2ascii
./build

${SUDO} rm -rf dist
mkdir dist

[ -d ${OUT_DIR} ] && rm -rf ${OUT_DIR}
mkdir ${OUT_DIR}
mkdir ${OUT_DIR}/DEBIAN
chmod 755 ${OUT_DIR} ${OUT_DIR}/DEBIAN

echo "Package: ${PKG}
Version: ${PKG_VER}-${PKG_REL}
Section: misc
Priority: optional
Architecture: ${ARCH}
Depends: curl, libjpeg-dev, libpng-dev, imagemagick
Maintainer: ${DEBFULLNAME} <${DEBEMAIL}>
Build-Depends: debhelper (>= 11)
Homepage: https://github.com/doctorfree/any2ascii
Description: Any2ascii converts any format image to an Ascii Art text file" > ${OUT_DIR}/DEBIAN/control

chmod 644 ${OUT_DIR}/DEBIAN/control

for dir in "${DESTDIR}" "${DESTDIR}/bin" "${DESTDIR}/share" \
           "${DESTDIR}/share/doc" "${DESTDIR}/share/man" \
           "${DESTDIR}/share/man/man1" "${DESTDIR}/share/man/man5" \
           "${DESTDIR}/share/doc/${PKG}" "${DESTDIR}/share/${PKG}"
do
    [ -d ${OUT_DIR}/${dir} ] || ${SUDO} mkdir ${OUT_DIR}/${dir}
    ${SUDO} chown root:root ${OUT_DIR}/${dir}
done

# Install any2ascii
${SUDO} cp src/jp2a ${OUT_DIR}/${DESTDIR}/bin/jp2a
${SUDO} cp src/any2a ${OUT_DIR}/${DESTDIR}/bin/any2a
${SUDO} cp man/jp2a.1 ${OUT_DIR}/${DESTDIR}/share/man/man1

${SUDO} cp AUTHORS ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}/AUTHORS
${SUDO} cp BUGS ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}/BUGS
${SUDO} cp COPYING ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}/COPYING
${SUDO} cp README ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}/README
${SUDO} cp copyright ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}
${SUDO} cp ChangeLog ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}
${SUDO} cp README.md ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}
${SUDO} cp LICENSE ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}
${SUDO} cp NEWS ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}
${SUDO} cp VERSION ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}
${SUDO} pandoc -f gfm README.md | ${SUDO} tee ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}/README.html > /dev/null
${SUDO} gzip -9 ${OUT_DIR}/${DESTDIR}/share/doc/${PKG}/ChangeLog

${SUDO} chmod 755 ${OUT_DIR}/${DESTDIR}/bin/* ${OUT_DIR}/${DESTDIR}/bin
${SUDO} chown -R root:root ${OUT_DIR}/${DESTDIR}

cd dist
echo "Building ${PKG_NAME}_${PKG_VER} Debian package"
${SUDO} dpkg --build ${PKG_NAME}_${PKG_VER} ${PKG_NAME}_${PKG_VER}-${PKG_REL}.${ARCH}.deb
cd ${PKG_NAME}_${PKG_VER}
echo "Creating compressed tar archive of ${PKG_NAME} ${PKG_VER} distribution"
tar cf - usr | gzip -9 > ../${PKG_NAME}_${PKG_VER}-${PKG_REL}.tgz

have_zip=`type -p zip`
[ "${have_zip}" ] || {
  ${SUDO} apt-get update
  ${SUDO} apt-get install zip -y
}
echo "Creating zip archive of ${PKG_NAME} ${PKG_VER} distribution"
zip -q -r ../${PKG_NAME}_${PKG_VER}-${PKG_REL}.zip usr

cd ..
[ "${GCI}" ] || {
    [ -d ../releases ] || mkdir ../releases
    [ -d ../releases/${PKG_VER} ] || mkdir ../releases/${PKG_VER}
    ${SUDO} cp *.deb *.tgz *.zip ../releases/${PKG_VER}
}
